Find the latest autoloader on first load.

zinigor opened the pull request:
This PR modifies the autoloader to find the latest version of itself among installed and active plugins and loads it.

*Note*: the commits are actually verified by precommit checks, they are like that because I had to rebase.

#### Changes proposed in this Pull Request:
* Moves the includable code to a separate file.
* Uses composer suffix for namespacing includable generated files to avoid conditionals and allow multiple autoloader versions to work together.
* Introduces the logic that finds the latest version available among *new* autoloaders and gives control to it.
* Fixes unit tests.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This is a Hack Week project to make autoloader work without waiting for all plugins to load.

#### Testing instructions:
* Use this branch, install Jetpack as usual.
* Use it with other plugins that use old autoloader, like VaultPress.
* Make a skeleton plugin that requires the new autoloader, make sure they work together fine, as long as Jetpack's autoloader works first. To make skeleton you can use a package.json file like this:
```
{
    "name": "zinigor/autoloader-test",
    "description": "An autoloader test",
    "type": "project",
    "require": {
        "automattic/jetpack-connection": "^1.9",
        "automattic/jetpack-autoloader": "dev-master"
    },
    "license": "GPL-2-or-later",
    "authors": [
        {
            "name": "Igor Zinovyev",
            "email": "zinigor@gmail.com"
        }
    ],
    "repositories": [
	{
	    "type": "path",
	    "url": "./packages/*"
	}
    ],
    "minimum-stability": "dev"
}
```
And here's the plugin file:
```
<?php
/**
 * Plugin Name: Autoloader test for Jetpack
 * Plugin URI: https://jetpack.com/support
 * Description: Bring the tests!
 * Author: Automattic
 * Version: 0.0.1-alpha
 * Author URI: https://jetpack.com
 * License: GPL2+
 * Text Domain: autoloader-test
 * Domain Path: /languages/
 *
 * @package Autoloader_Test
 */

require_once trailingslashit( dirname( __FILE__ ) ) . 'vendor/autoload_packages.php';

function instantiate_a_bunch_of_classes() {
  //new IXR_Client();
  //new Jetpack_IXR_Client();
  //new Jetpack_Options();
}

instantiate_a_bunch_of_classes();
```
Notice that you will need to copy (or link) the packages folder from Jetpack in order to test the new version of the autoloader.
* You should have no problem running Jetpack alongside this skeleton plugin, but as soon as you tinker with the version in the classmap to make the skeleton's autoloader in charge, you should get a fatal error about classes not being found. That part is still in progress.

#### Proposed changelog entry for your changes:
N/A
*

kbrown9 committed:
[not verified] Autoloader: add the set_up_autoloader function.

kbrown9 committed:
[not verified] Autoloader: update global array names and remove spl_autoload_register

Update the global classmap and filemap array names so the old autoloaders
can't overwrite the new autoloader's version selections.

Remove the call to spl_autoload_register(), because that it now handled
in set_up_autoloader().

zinigor committed:
[not verified] Moved the autoloader functions into a separate include.

kbrown9 commented on packages/autoloader/tests/php/test_file_loader.php:
```diff
@@ -1,66 +1,80 @@
-		// re-apply the global
-		global $jetpack_packages_files;
-		$jetpack_packages_files = self::$jetpack_packages_files;
+		spl_autoload_register( 'autoloader' );
```
I think we can remove this `spl_autoload_register()` call because these tests just test the 'autoloading' of files. The spl autoloader just autoloads classes.

kbrown9 committed:
Autoloader: remove setup() from test_file_loader.php

The spl_autoload_register() call isn't needed in setup() and the
setup() method doesn't do anything else, so remove the entire setup()
method.

kbrown9 reviewed (approved):
This is working for me! I tested with multiple plugins with the old and new autoloader. The new autoloader successfully identifies the latest installed autoloader.

@zinigor - please review commit bfd3d2c and merge into the feature branch if you approve.
